### ISSUE TITLE
Exception doesent kill job with ex-handling flow conditions

### ISSUE THREAD

**Issue by gardnervickers** [onyx-platform/onyx#272]
When a job has flow conditions that handle exceptions, any task in the workflow that is not covered by the ex-handling logic can hang the job if it throw's and exception. Reproduced on this branch
https://github.com/gardnervickers/onyx/tree/flow_exception. All that was done is add another task upstream that throws an exception. 

Log entry's are being created and successfully pushed onto `outbox-ch` inside of `onyx.peer.task-lifecycle/handle-exception`. Something is not letting the job die.

**Comment by gardnervickers** [comment]
If you don't call `(take-segments! out-chan)` then it terminates. Somewhere blocking on the out-chan is hanging the job shutdown.

**Comment by dignati** [comment]
We seem to see a similar issue. Some tasks in our workflow are covered by an exception handler and flow-conditions are put into place to handle those. Jobs seem to finish but `onyx/await-job-completion` never returns.

**Comment by lbradstreet** [comment]
This is because the job is being killed, and await-job-completion only returns when a job is completed not killed. I suggest the following enhancement to await-job-completion, which will allow await-job-completion to return when a job is killed or completed, and use the return value to denote success or failure (killing).

https://github.com/onyx-platform/onyx/commit/be870d137f3ff255dd842749c6537c3ac1b83fc1

@MichaelDrogalis, thoughts?

**Comment by MichaelDrogalis** [comment]
Absolutely agreed to the change here. Thanks for tracking this one down!

**Comment by lbradstreet** [comment]
Merged into master, shipped a snapshot out at 0.7.1-SNAPSHOT. Waiting to hear back from @gardnervickers before closing.

**Comment by gardnervickers** [comment]
Turns out I had this wrong. I was under the assumption that `:done` was supposed to be sent regardless, signifying that a job had finished or died. Makes a lot more sense that it is supposed to be sent only on successful completion. 

I was expecting done to unblock `(take-segments! out-chan)` no matter what, signifying that a job was successfully killed. 

If you go back to my [branch](https://github.com/gardnervickers/onyx/tree/), you can comment out the flow conditions [here](https://github.com/gardnervickers/onyx/blob/flow_exception/test/onyx/peer/flow_exception_test.clj#L131).

Run the test and you can check the results seq. It contains exceptions for each of the segments that triggered the `(throw)`'s, and finally it contains `:done`. 

`:done` gets passed, and the job continues, even if a task throws an exception.

**Comment by MichaelDrogalis** [comment]
Fixed by #274.
